### PR TITLE
More workspaces tests

### LINKED ISSUE #1154: Workspaces Test Suite
We keep changing the Workspaces API, and yet we don't have a proper test suite for it! We should change this....pronto.

### PR DESCRIPTION
**Pull request by greenfrogs**
## What does this PR do?
Adds tests for all workspaces cache components. Closes #1154

## ✅ Pull Request checklist

- [x] Is this code complete?
- [x] Are tests done/modified?

## 🚨 Is this a breaking change for API clients?
No

### CODE REVIEW

File: backend/uclapi/workspaces/views.py [hunk 1]
```diff
@@ -1,82 +1,63 @@
...
-            "ok": False,
-            "error": (
-                "The image with the ID you requested "
-                "does not exist."
-            )
-        }, custom_header_data=kwargs)
+        response = JsonResponse(
+            {
+                "ok": False,
+                "error": ("The image with the ID you requested " "does not exist."),
```
Comment (ILW8) [inline backend/uclapi/workspaces/views.py:57]: can drop parentheses and two quotes

File: backend/uclapi/workspaces/views.py [hunk 2]
```diff
@@ -86,342 +67,289 @@ def get_map_image(request, *args, **kwargs):
...
-            "ok": False,
-            "error": (
-                "You specified a response format that "
-                "was not either raw or base64."
-            )
-        }, custom_header_data=kwargs)
+        response = JsonResponse(
+            {
+                "ok": False,
+                "error": ("You specified a response format that " "was not either raw or base64."),
```
Comment (ILW8) [inline backend/uclapi/workspaces/views.py:82]: same as the above comment

File: backend/uclapi/workspaces/views.py [hunk 3]
```diff
@@ -1,82 +1,63 @@
...
-            "ok": False,
-            "error": (
-                "The survey filter you provided is invalid. "
-                "Valid survey filters are: "
-            ) + str(consts.VALID_SURVEY_FILTERS)
-        }, custom_header_data=kwargs)
+        response = JsonResponse(
+            {
+                "ok": False,
+                "error": ("The survey filter you provided is invalid. " "Valid survey filters are: ")
```
Comment (ILW8) [inline backend/uclapi/workspaces/views.py:27]: can drop parentheses and two quotes

File: backend/uclapi/workspaces/views.py [hunk 4]
```diff
@@ -86,342 +67,289 @@ def get_map_image(request, *args, **kwargs):
...
```
Comment (ILW8) [inline backend/uclapi/workspaces/views.py:149]: same as above

File: backend/uclapi/workspaces/views.py [hunk 5]
```diff
@@ -86,342 +67,289 @@ def get_map_image(request, *args, **kwargs):
...
```
Comment (ILW8) [inline backend/uclapi/workspaces/views.py:188]: same as above

**Review by ILW8 — COMMENTED** [review]
Looks good overall. One note about adding cur_time to `cache_historical_time_usage_data`, otherwise only small formatting issues in `views.py`

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 6]
```diff
@@ -333,53 +212,27 @@ def cache_sensors_for_map(self, survey_id, map_id):
...
+        map_vmax_y_key = self._const.SURVEY_MAP_VMAX_Y_KEY.format(survey_id, map_id)
+        map_viewbox_key = self._const.SURVEY_MAP_VIEWBOX_KEY.format(survey_id, map_id)
+        pipeline.set(map_vmax_x_key, all_map_sensors_data["VMaxX"])
+        pipeline.set(map_vmax_y_key, all_map_sensors_data["VMaxY"])
+        pipeline.set(map_viewbox_key, all_map_sensors_data["ViewBox"])
 
         pipeline.execute()
 
-    def cache_historical_time_usage_data(self, survey_id, day_count):
+    def cache_historical_time_usage_data(self, survey_id, day_count, cur_date=datetime.now()):
```
[the remark this answers is not in the document]
Comment (greenfrogs) [reply backend/uclapi/workspaces/occupeye/cache.py:235]: Yes we could and using freezegun makes for better tests. But unless we are using freezegun elsewhere, this solution does not change the function signature. cur_date is completely optional. However saying all of that it prevents the testing of cache_historical_time_usage_data from other functions (for example feed_cache) so we probably want to change it.

File: backend/uclapi/workspaces/views.py [hunk 7]
```diff
@@ -86,342 +67,289 @@ def get_map_image(request, *args, **kwargs):
...
-            "error": (
-                "You did not specify how many days of historical data "
-                "should be returned. Valid options are: "
-            ) + str(consts.VALID_HISTORICAL_DATA_DAYS)
-        }, custom_header_data=kwargs)
+        response = JsonResponse(
+            {
+                "ok": False,
+                "error": (
+                    "You did not specify how many days of historical data " "should be returned. Valid options are: "
```
[the remark this answers is not in the document]
Comment (greenfrogs) [reply backend/uclapi/workspaces/views.py:203]: Thank you

File: backend/uclapi/workspaces/occupeye/cache.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ILW8) [reply backend/uclapi/workspaces/occupeye/cache.py:235]: It does look like freezegun is [being used elsewhere](https://github.com/uclapi/uclapi/blob/1a9bfd2fd3cbbf213b2ae360d57939af553ca08d/backend/uclapi/common/tests.py#L30)!

[the remark this answers is not in the document]
Comment (greenfrogs) [reply backend/uclapi/workspaces/occupeye/cache.py:235]: Updated!

**Review by ILW8 — DISMISSED** [review]
Looks good

**Review by zipy124 — CHANGES_REQUESTED** [review]
See attached comment. Apart from that absolutely beautiful, ready to get to master when the small change is implemented (or you decide isn't needed :shrug: )

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py [hunk 8]
```diff
@@ -1,53 +1,59 @@
...
                 except CachetException as cachet_error:
-                    print(f"Failed to create cachet incident. "
-                          f"Reason: {repr(cachet_error)}")
+                    print(f"Failed to create cachet incident. " f"Reason: {repr(cachet_error)}")
                 except Exception as cachet_error:
-                    print(f"Unexpected: Failed to create cachet incident. "
-                          f"Reason: {repr(cachet_error)}")
+                    print(f"Unexpected: Failed to create cachet incident. " f"Reason: {repr(cachet_error)}")
             else:
+                import sys
```
[the remark this answers is not in the document]
Comment (greenfrogs) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py]: I agree, this should be at the top of the file.

**Comment by ILW8** [comment]
Looks even better 👍

**Pull request merged**